### ISSUE TITLE
Sequential creation of related columns using apply-operation command

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#500]
_Original author: gofranshukair@gmail.com (November 29, 2011 22:14:02)_

<b>What steps will reproduce the problem?</b>
1.applying add-columns operations using json file 

<b>What is the expected output? What do you see instead?</b>

<b>What version of Google Refine are you using?</b>
2.5
<b>What operating system and browser are you using?</b>
Ubuntu 
<b>Is this problem specific to the type of browser you're using or it happens</b>
<b>in all the browsers you tried?</b>
not browser-specific

<b>Please provide any additional information below.</b>

I am working on Google refine extension that requires applying a number of grefine json operations after creating the project .
I implemented a custom apply-operations command  which will be called after launching the /project page.
Basically the json operations are created manually on Google refine then I extracted the history into a file that and I am reading in the command doPost function and apply them to the project as the following :

```
public void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        Project project = getProject(request);
        String jsonString = getOperationFile();

    try {
        JSONArray a = ParsingUtilities
                .evaluateJsonStringToArray(jsonString);
        int count = a.length();
        for (int i = 0; i &lt; count; i++) {
            JSONObject obj = a.getJSONObject(i);
            reconstructOperation(project, obj);
        }

        if (project.processManager.hasPending()) {
            respond(response, &quot;{ \&quot;code\&quot; : \&quot;pending\&quot; }&quot;);
        } else {
            respond(response, &quot;{ \&quot;code\&quot; : \&quot;ok\&quot; }&quot;);
        }
    } catch (JSONException e) {
        respondException(response, e);
    }
}
```

The file includes &quot;core/column-addition-by-fetching-urls&quot; operations that receive json data , then other operations create columns from the splitting that json data in the previously created columns 
The problem that : The function &quot;createProcess&quot; of the  &quot;ColumnAdditionByFetchingURLsOperation&quot; checks if baseColumnName of the operation exists in the project.columnModel before creating it, and throw an exception.
That is happening because the creation process of the _baseColumnName is still pending or not fully executed. 

```
public Process createProcess(Project project, Properties options) throws Exception {
        Column column = project.columnModel.getColumnByName(_baseColumnName);
        if (column == null) {
            throw new Exception(&quot;No column named &quot; + _baseColumnName);
        }
        if (project.columnModel.getColumnByName(_newColumnName) != null) {
            throw new Exception(&quot;Another column already named &quot; + _newColumnName);
        }

    Engine engine = createEngine(project);
    engine.initializeFromJSON(_engineConfig);

    Evaluable eval = MetaParser.parse(_urlExpression);

    return new ColumnAdditionByFetchingURLsProcess(
        project, 
        engine,
        eval,
        getBriefDescription(null)
    );
}
```

I understand that related columns operations should be executed sequentially as their order in the json file.

_Original issue: http://code.google.com/p/google-refine/issues/detail?id=500_

**Comment by thadguidry** [comment]
This needs review to see if the bug indeed still does exist.  Someone should just write a test case around this issue and then fix if the issue comes up.  @ostephens wanna hack on this and take your OpenRefine coding skills to the next level ? :)

**Comment by wetneb** [comment]
This is potentially related to the way we handle the queue of pending operations, that I plan to work on for #2256.

**Comment by tfmorris** [comment]
I've fixed the code formatting in the original issue that apparently got messed up during the transfer from Google Code.

David's attempted fix in r2394 (aka a7e27046550493c64b493076d6c14d098e2c9646) moved, but apparently didn't resolve, the problem.

Although fancy parallelization like #2256 could be done, it seems like a tweak to the execution scheme could close the race window, perhaps at some small performance cost, which could be optionally incurred, like a sync point that would wait for operations to drain.

Of course, since @gofranshukair has long probably long since lost interest, and the internal client/server protocol isn't documented for external use, I'd be tempted to close this unless the problem manifests itself in a manual "Apply" of an operation history containing multiple Add Column commands which have dependencies on their predecessor.

**Comment by wetneb** [comment]
This is solved in 4.0, since with partial result support we update the column model before the operation data is fetched.

**Comment by tfmorris** [comment]
As part of improving the error handling for `apply-operations`, the test described above:

> "Apply" of an operation history containing multiple Add Column commands which have dependencies on their predecessor.

would be useful to confirm whether or not the problem still exists.
